File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    static String[] key = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
    public List<String> letterCombinations(String digits) {
        List<String>ll= new ArrayList<>();
        if(digits.length()==0){
            return ll;
        }
        phonenumber(digits,"",ll);
        return ll;
    }
    public static void phonenumber(String ques,String ans,List<String> ll){
        if(ques.length()==0){
            ll.add(ans);
            return;
        }
        char ch=ques.charAt(0);
        String getstring=key[ch-'0'];
        for(int i=0;i<getstring.length();i++){
            phonenumber(ques.substring(1),ans+getstring.charAt(i),ll);
        }
    }

}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> ll;
    vector<string> letterCombinations(string digits) {
        string val[] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        string t="";
        solve(digits,0,t,val);
        return ll;
    }
    void solve(string &s,int idx,string ans,string val[]){
        if(idx==s.length()){
            ll.push_back(ans);
            return;
        }
        int d=s[idx]-'0';
        for(int i=0;i<val[d].size();i++){
            solve(s,idx+1,ans+val[d][i],val);
        }
    }
};